WalletWasabi/Blockchain/TransactionOutputs/CoinsRegistry.cs

```diff
@@ -230,57 +148,128 @@ public bool IsKnown(uint256 txid)
 		}
 	}
 
-	private bool TryGetSpenderSmartCoinsByOutPointNoLock(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
+	private bool TryGetCoinsByInputPrevOutNoLock(OutPoint outPoint, [NotNullWhen(true)] out HashSet<SmartCoin>? coins)
 	{
-		return CoinsByOutPoint.TryGetValue(outPoint, out coins);
+		coins = null;
+		lock (Lock)
+		{
+			if (!TxIdsByInputsPrevOut.TryGetValue(outPoint, out var txId))
+			{
+				return false;
+			}
+
+			if (!CoinsByTransactionsIds.TryGetValue(txId, out var coinsByOutpoints))
+			{
+				return false;
+			}
+
+			coins = coinsByOutpoints.Values.ToHashSet();
+		}
+		return true;
 	}
 
-	public bool TryGetSpentCoinByOutPoint(OutPoint outPoint, [NotNullWhen(true)] out SmartCoin? coin)
+	public bool TryGetCoinByOutPoint(OutPoint outPoint, [NotNullWhen(true)] out SmartCoin? coin)
 	{
+		coin = null;
 		lock (Lock)
 		{
-			return SpentCoinsByOutPoint.TryGetValue(outPoint, out coin);
+			if (!CoinsByTransactionsIds.TryGetValue(outPoint.Hash, out var coinsByOutpoint))
+			{
+				return false;
+			}
+
+			if (!coinsByOutpoint.TryGetValue(outPoint, out coin))
+			{
+				return false;
+			}
 		}
+
+		return true;
 	}
 
 	internal (ICoinsView toRemove, ICoinsView toAdd) Undo(uint256 txId)
 	{
 		lock (Lock)
 		{
 			var allCoins = AsAllCoinsViewNoLock();
-			var toRemove = new List<SmartCoin>();
-			var toAdd = new List<SmartCoin>();
+			var toRemove = new HashSet<SmartCoin>();
+			var toAdd = new HashSet<SmartCoin>();
 
-			// remove recursively the coins created by the transaction
-			foreach (SmartCoin createdCoin in allCoins.CreatedBy(txId))
+			// The transactions that needs to be removed are txId and its descendants.
+			HashSet<uint256> txIdsToRemove = new () { txId };
+			foreach (var myOutput in allCoins.CreatedBy(txId))
 			{
-				toRemove.AddRange(RemoveNoLock(createdCoin));
+				var txIdsToRemoveForOutput = allCoins
+					.DescendantOf(myOutput)
+					.Select(x => x.TransactionId)
+					.Distinct();
+
+				foreach (var txIdToRemove in txIdsToRemoveForOutput)
+				{
+					txIdsToRemove.Add(txIdToRemove);
+				}
 			}
 
-			// destroyed (spent) coins are now (unspent)
-			foreach (SmartCoin destroyedCoin in allCoins.SpentBy(txId))
+			foreach (var txIdToRemove in txIdsToRemove)
 			{
-				if (SpentCoins.Remove(destroyedCoin))
+				// Remove the transaction from CoinsByTransactionIds cache.
+				if (CoinsByTransactionsIds.Remove(txIdToRemove, out var coinsRemovedByOutpoint))
+				{
+					foreach (var removedCoin in coinsRemovedByOutpoint.Values)
+					{
+						toRemove.Add(removedCoin);
+					}
+
+					// Remove the coins from the CoinsByPubKeys cache.
+					foreach (var coinsByPubKey in CoinsByPubKeys.ToList())
+					{
+						if (coinsByPubKey.Value.Count == coinsByPubKey.Value.RemoveWhere(x => coinsRemovedByOutpoint.ContainsValue(x)))
+						{
+							// There are no more coins on the PubKey, remove the entry.
+							CoinsByPubKeys.Remove(coinsByPubKey.Key);
+						}
+					}
```

Comment:
Question just to be sure:

Here we are iterating over `CoinsByPubKeys` and also removing from it.

Is this OK because we materialize it with `CoinsByPubKeys.ToList()`, right?

Reply:
Yes